src/differential_arithmetic.jl

```diff
@@ -88,6 +92,8 @@ for T in (:AbstractThunk, :Composite, :Any)
     @eval Base.:*(a::$T, ::One) = a
 end
 
+LinearAlgebra.dot(::One, x::Number) = x
+LinearAlgebra.dot(x::Number, ::One) = conj(x)
```

Comment:
Why these overloads? Because `dot` with `One` also needs its recursion broken (see example in the OP).
Why conjugate? Because `dot(x::Number, 1) == conj(x)`.
Why limit to numbers? Because I have no idea what a user means by `dot(One(), [1.0, 2.0, 3.0])`.
Or have I misunderstood?

Reply:
> Why conjugate? Because dot(x::Number, 1) == conj(x).

that is surprising/interesting to me.

Reply:
> > Why conjugate? Because dot(x::Number, 1) == conj(x).
> 
> that is surprising/interesting to me.

It's not documented as such, but `dot` in Base implements a [Frobenius inner product](https://en.wikipedia.org/wiki/Frobenius_inner_product) that is conjugate linear in the first argument. Other packages (can't think of any off the top of my head, but there were arguments on slack about this) extend `dot` for other kinds of inner products, but I think in general they use different signatures.

Reply:
```suggestion
LinearAlgebra.dot(x::Number, ::One) = conj(x)  # see definition of Frobenius inner product
```